### PR TITLE
infoschema: fix issue of information schema cache miss cause by schema version gap

### LINKED ISSUE #53428: stale-read query latency 10x spike cause by information schema cache miss
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

High schema cache miss.

Related metrics:

image

image

#### Analyze
    The reason for a large number of load snapshot schema operations must be InfoSchema cache miss, but what causes cache miss?

    Almost all TiDB generate a large number of load snapshot schema operations at the same time. It's weird, I picked a TiDB (tidb-11) with a large amount of operations for log analysis.

> grep 'load InfoSchema' tidb-11.log

```log
"[2024/05/09 15:00:46.337 +00:00] [INFO] [domain.go:224] [\"diff load InfoSchema success\"] [currentSchemaVersion=585441] [neededSchemaVersion=585442] [\"start time\"=1.734944ms] [phyTblIDs=\"[]\"] [actionTypes=\"[]\"]",

"[2024/05/09 15:00:46.521 +00:00] [INFO] [domain.go:224] [\"diff load InfoSchema success\"] [currentSchemaVersion=585442] [neededSchemaVersion=585443] [\"start time\"=503.56µs] [phyTblIDs=\"[]\"] [actionTypes=\"[]\"]",

"[2024/05/09 15:01:34.130 +00:00] [INFO] [domain.go:224] [\"diff load InfoSchema success\"] [currentSchemaVersion=585443] [neededSchemaVersion=585445] [\"start time\"=4.158976ms] [phyTblIDs=\"[454637]\"] [actionTypes=\"[8]\"]",
```

The schema version has 585442, 585443, and 585445, but 585444 does not exist, and there's a hole that causes schema cache misses, so TiDB needs to load snapshot schema from TiKV. Related codes are following:

https://github.com/pingcap/tidb/blob/868456d0f2b89e032997c3c07cad633725583e67/infoschema/cache.go#L118

Well, Why does the schema version 585444 not exist? From the current code implementation, since `GenSchemaVersion` uses a separate transaction to generate schema version, but `SetSchemaDiff` use another transaction, so if `GenSchemaVersion` transaction succeed, but `SetSchemaDiff` transaction fails,  then will have schema version gap in this issue.

Why use two separate transactions? this change is introduced by [PR: support concurrent ddl](https://github.com/pingcap/tidb/pull/32169/files#diff-dfc42c5764e7e4ff9122a1db728ff1cb0dee56e72dbccefdb211018ccd444c73L1131)

https://github.com/pingcap/tidb/blob/5d990c60e4a2928148cde450cd37b38acded2724/pkg/ddl/ddl.go#L448-L453

https://github.com/pingcap/tidb/blob/851e22d0b2b221388a076115ed76720c27b286cf/pkg/ddl/schema_version.go#L358

https://github.com/pingcap/tidb/blob/2abf83ddd50755dc7c3af6d2f2cf669fab29051d/domain/domain.go#L423-L424

### 2. What did you expect to see? (Required)

No information schema cache miss.

### 3. What did you see instead (Required)

information schema cache miss then cause many operation which is load snapshot schema for stale-read query.

### 4. What is your TiDB version? (Required)

nightly: e1a6b1d63399ce06b00132bc55052fb2c4e8a7fd

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Issue Number: close #53428

Problem Summary: fix issue of information schema cache miss cause by schema version gap

#### before this PR

When DDL has empty schema version, will cause cache miss, then stale-read query need to load snapshot schema from TiKV.

image

```sql
[2024/05/23 06:15:10.546 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6760] [neededSchemaVersion=6762] ["start time"=2.583049ms] [gotSchemaVersion=6762] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:15:17.869 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6763] [neededSchemaVersion=6765] ["start time"=2.040666ms] [gotSchemaVersion=6765] [phyTblIDs="[10496]"] [actionTypes="[4]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:15:18.051 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6765] [neededSchemaVersion=6767] ["start time"=4.321375ms] [gotSchemaVersion=6767] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:15:18.245 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6767] [neededSchemaVersion=6769] ["start time"=3.01603ms] [gotSchemaVersion=6769] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
```

### This PR

When DDL has empty schema version, this PR won't have schema cache miss, then won't cause many load snapshot schema operation.

image

```log
[2024/05/23 06:23:45.209 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6777] [neededSchemaVersion=6779] ["start time"=5.409896ms] [gotSchemaVersion=6779] [phyTblIDs="[10502]"] [actionTypes="[4]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:45.458 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6779] [neededSchemaVersion=6781] ["start time"=2.711853ms] [gotSchemaVersion=6781] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:45.650 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6781] [neededSchemaVersion=6783] ["start time"=1.910751ms] [gotSchemaVersion=6783] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:49.173 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6784] [neededSchemaVersion=6786] ["start time"=2.511772ms] [gotSchemaVersion=6786] [phyTblIDs="[10505]"] [actionTypes="[4]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:49.372 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6786] [neededSchemaVersion=6788] ["start time"=2.768253ms] [gotSchemaVersion=6788] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:49.590 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6788] [neededSchemaVersion=6790] ["start time"=4.466332ms] [gotSchemaVersion=6790] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:53.670 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6791] [neededSchemaVersion=6793] ["start time"=799.389µs] [gotSchemaVersion=6793] [phyTblIDs="[10508]"] [actionTypes="[4]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:53.872 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6793] [neededSchemaVersion=6795] ["start time"=1.025364ms] [gotSchemaVersion=6795] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
[2024/05/23 06:23:54.090 +00:00] [INFO] [domain.go:287] ["diff load InfoSchema success"] [currentSchemaVersion=6795] [neededSchemaVersion=6797] ["start time"=3.549838ms] [gotSchemaVersion=6797] [phyTblIDs="[]"] [actionTypes="[]"] [diffTypes="[\"drop table\"]"]
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by crazycs520** [comment]
@lcwangchao @zimulala @tangenta PTAL

File: pkg/infoschema/cache.go [hunk 1]
```diff
@@ -241,3 +271,24 @@ func (h *InfoCache) Insert(is InfoSchema, schemaTS uint64) bool {
...
+
+	h.emptySchemaVersions[version] = struct{}{}
+	if len(h.emptySchemaVersions) > cap(h.cache) {
+		// remove oldest version.
+		versions := make([]int64, 0, len(h.emptySchemaVersions))
+		for ver := range h.emptySchemaVersions {
+			versions = append(versions, ver)
+		}
+		sort.Slice(versions, func(i, j int) bool { return versions[i] < versions[j] })
+		for _, ver := range versions {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply pkg/infoschema/cache.go:287]: Because empty schema version rarely appears, I think it's okay to delete just one here.

**Review by lcwangchao — APPROVED** [review]

File: pkg/infoschema/cache.go [hunk 2]
```diff
@@ -241,3 +271,25 @@ func (h *InfoCache) Insert(is InfoSchema, schemaTS uint64) bool {
...
 	return true
 }
+
+// InsertEmptySchemaVersion inserts empty schema version into a map. If exceeded the cache capacity, remove the oldest version.
+func (h *InfoCache) InsertEmptySchemaVersion(version int64) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	h.emptySchemaVersions[version] = struct{}{}
+	if len(h.emptySchemaVersions) > cap(h.cache) {
```
Comment (wjhuang2016) [inline pkg/infoschema/cache.go:281]: @tiancaiamao Is it ok to use `cap` here?

File: pkg/infoschema/cache.go [hunk 3]
```diff
@@ -241,3 +271,25 @@ func (h *InfoCache) Insert(is InfoSchema, schemaTS uint64) bool {
...
+		// remove oldest version.
```
Comment (D3Hunter) [inline pkg/infoschema/cache.go:282]: without statistics of how those version are accessed, remove oldest is no better than remove random version, and more complex

Comment (crazycs520) [reply pkg/infoschema/cache.go:282]: For stale-read query workload, it will always need  recent schema version, so remove the oldest version is better.

**Review by tiancaiamao — COMMENTED** [review]
How about insert a clone of previous infoschema for the empty schema diff? with its schema version = previous schema version + 1
The data structure and all the other operations can be the same as before.

**Review by tiancaiamao — APPROVED** [review]

**Comment by crazycs520** [comment]
> How about insert a clone of previous infoschema for the empty schema diff? with its schema version = previous schema version + 1 The data structure and all the other operations can be the same as before.

Nice catch, I try to use methold, but this will need to introduce another TiKV RPC to get the schemaTS of the empty schema version.

**Comment by crazycs520** [comment]
@wjhuang2016 @D3Hunter PTAL

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**